src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -80,9 +80,112 @@ internal static T[] ToArray<T>(IEnumerable<T> source)
                 return result;
             }
 
-            LargeArrayBuilder<T> builder = new();
```

Comment:
Thanks. What would it take to augment your PR to the point where we could delete LargeArrayBuilder?  That'd make me much more interested in this.

Reply:
LargeArrayBuilder is used in many places, and given that it includes numerous methods specific to certain scenarios, the amount of rewrites, including from the user's side, would likely be extensive. Ideally, we'd like to replace it entirely, but we want to exclude that from the initial implementation.

Using the ArrayPool as suggested in the option might not be advisable if we replace the LargeArrayBuilder.

Reply:
My concern is that these code paths are used in a bunch of places on common paths, and adding more generic types will increase code size. It'd be nice to get rid of the LargeArrayBuilder so that it's a net win for both throughput and size, rather than improving throughput at the expense of size.

There's no rush here as it's not going to make .NET 8, anyway, so there's plenty of time to make and evaluate the larger change.

Reply:
Thank you.
If this implementation is acceptable as a basic policy,
I'll take on the challenge of replacing the LargeArrayBuilder.
(Indeed, I was hoping it might make it into .NET 8! That would be great!)

Reply:
> If this implementation is acceptable as a basic policy,

We'd support it in principle. Would you like to build on top of this PR or would you prefer to open a new one?

Reply:
I am willing to work on this PR.
NET 8 release seems to be coming soon, so I will do it from the branch there when it is released. ......